### PR TITLE
feat: add kernel conflict evaluators

### PR DESCRIPTION
**Pull request by harshanandak**
## Summary
- Add Kernel conflict evaluator contract for stale revision quarantine, idempotency replay, equivalent write dedupe, and dependency-cycle guards.
- Add broker guarded-event ordering so conflicts are inserted before projection and accepted events enqueue projection only after event insertion.
- Add evaluator fixtures and docs for the PR D migration UX handoff.

## Validation
- bun test test/kernel/evaluators.test.js test/kernel/evaluator-fixtures.test.js test/kernel/broker-guards.test.js test/kernel/broker.test.js test/kernel/schema.test.js test/kernel/conflict-evaluator-docs.test.js test/kernel-schema-docs.test.js
- bun run lint
- git diff --check --cached
- pre-push hook: lint, targeted tests, affected edge-case tests

## Summary by CodeRabbit

* **New Features**
  * Kernel conflict evaluators: detect/quarantine conflicting writes, dedupe equivalent writes, handle idempotency replays, detect dependency cycles, and ensure projection outbox rows are enqueued only after acceptance; guarded kernel event ingestion added.

* **Documentation**
  * Added reference, design, decisions, tasks and schema notes covering evaluator behavior, quarantines, idempotency, migrations, and test guidance.

* **Tests**
  * New fixtures and suites validating evaluator outcomes, guard ordering, idempotency, dedupe, cycle and stale-revision handling, plus docs coverage checks.

* **Chores**
  * Schema, migrations, and adapter imports updated to include expected_revision and migration plan adjustments.

### CODE REVIEW

File: lib/kernel/broker.js [hunk 1]
```diff
@@ -106,6 +117,68 @@ function createLocalBroker(options = {}) {
...
+
+			const config = getConfig();
+			const now = context.now || new Date().toISOString();
+			const normalizedEvent = {
+				...event,
+				created_at: event.created_at || now,
+			};
+			const [entity, priorEvents, dependencies] = await Promise.all([
+				driver.loadKernelEntity(normalizedEvent.entity_type, normalizedEvent.entity_id, context, config),
+				driver.listKernelEvents(normalizedEvent.entity_type, normalizedEvent.entity_id, context, config),
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/broker.js:192]: Fixed in f43e8b3: runGuardedEvent now checks idempotency globally via loadKernelEventByIdempotencyKey before insert/projection. Focused tests and pre-push passed.

File: lib/kernel/evaluators.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+function dependencyCreatesCycle(payload, dependencies = []) {
+	const issueId = payload.issue_id;
+	const blocksIssueId = payload.blocks_issue_id;
+	if (!issueId || !blocksIssueId) return false;
+
+	const edges = new Map();
+	for (const dependency of [...dependencies, { issue_id: issueId, blocks_issue_id: blocksIssueId }]) {
+		if (!edges.has(dependency.issue_id)) edges.set(dependency.issue_id, []);
+		edges.get(dependency.issue_id).push(dependency.blocks_issue_id);
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/evaluators.js]: Fixed in f43e8b3: cycle detection now ignores non-blocking dependency types such as related and parent-child while preserving blocks/legacy blocking semantics. Added direct tests and pre-push passed.

File: lib/kernel/evaluators.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+function findEquivalentEvent(event, priorEvents = []) {
+	const payloadJson = payloadJsonFor(event);
+	return priorEvents.find(priorEvent => (
+		priorEvent.idempotency_key !== event.idempotency_key
+		&& priorEvent.entity_type === event.entity_type
+		&& priorEvent.entity_id === event.entity_id
+		&& priorEvent.event_type === event.event_type
+		&& payloadJsonFor(priorEvent) === payloadJson
+	));
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/evaluators.js:46]: Fixed in 34696c4: equivalent-write dedupe now requires true retry evidence by matching the same expected_revision, so an intentional later return to an older payload is accepted and projected. Added regression coverage for A -> B -> A and updated the fixture case. Focused tests and pre-push passed.

File: lib/kernel/evaluators.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+function findOriginalIdempotencyEvent(event, priorEvents = []) {
+	return priorEvents.find(priorEvent => priorEvent.idempotency_key === event.idempotency_key);
+}
+
+function findEquivalentEvent(event, priorEvents = []) {
+	const payloadJson = payloadJsonFor(event);
+	return priorEvents.find(priorEvent => (
+		priorEvent.idempotency_key !== event.idempotency_key
+		&& priorEvent.expected_revision !== undefined
+		&& Number(priorEvent.expected_revision) === Number(event.expected_revision || 0)
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/evaluators.js:41]: Fixed in 849f3a2: kernel_events now persists expected_revision, the migration expectation covers the column, accepted event records normalize expected_revision before storage, and the schema docs/tests mention why the metadata is required for equivalent retry dedupe. Focused tests and pre-push passed.

File: lib/kernel/schema.js [hunk 5]
```diff
@@ -196,6 +196,7 @@ const TABLE_LIST = deepFreeze([
 		field('entity_id', 'TEXT', { notNull: true }),
 		field('event_type', 'TEXT', { notNull: true }),
 		field('idempotency_key', 'TEXT', { notNull: true }),
+		field('expected_revision', 'INTEGER', { notNull: true }),
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/schema.js:199]: Fixed in 92945fa: added additive migration 002_kernel_events_expected_revision so existing Kernel DBs gain kernel_events.expected_revision via ALTER TABLE, while 001_kernel_schema remains the historical base shape. Focused tests and pre-push validation passed.

[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/schema.js:199]: Fixed in 92945fa: Beads close-event imports now include expected_revision from issue.entity_revision with a 0 fallback, matching the required Kernel event schema. Adapter tests passed.

File: lib/kernel/broker.js [hunk 6]
```diff
@@ -67,6 +68,38 @@ function requireDriverMethod(driver, methodName) {
...
+function buildDependencyScope(event) {
+	if (event.event_type !== 'dependency.add') return null;
+	const payload = parseEventPayload(event);
+	return {
+		issue_id: payload.issue_id,
+		blocks_issue_id: payload.blocks_issue_id,
+		dependency_type: payload.dependency_type || 'blocks',
+		entity_type: event.entity_type,
+		entity_id: event.entity_id,
+	};
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/broker.js:121]: Fixed in 92945fa: buildDependencyScope() now returns null unless both dependency endpoints are present, so incomplete dependency drafts cannot trigger a broad dependency lookup. Added broker guard coverage.

File: lib/kernel/migrations.js [hunk 7]
```diff
@@ -79,6 +93,14 @@ function buildSchemaMigration(schema = getKernelSchema()) {
 	};
 }
 
+function buildEventExpectedRevisionMigration() {
+	return {
+		id: '002_kernel_events_expected_revision',
+		apply: ['ALTER TABLE kernel_events ADD COLUMN expected_revision INTEGER NOT NULL DEFAULT 0;'],
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/migrations.js:99]: Fixed in c59444e: broker initialization now guards the additive expected_revision column migration against duplicate-column replay, so repeated initialize() calls do not fail after the first successful startup. Added broker regression coverage and pre-push passed.

File: lib/adapters/beads-kernel-compat.js [hunk 8]
```diff
@@ -347,6 +347,7 @@ function buildCloseEvent(issue, importedAt) {
     entity_id: issue.id,
     event_type: 'beads.issue.closed',
     idempotency_key: `beads-close:${issue.id}:${createdAt}`,
+    expected_revision: Number(issue.entity_revision || 0),
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/adapters/beads-kernel-compat.js]: Fixed in c59444e: Beads close-event import now normalizes entity_revision with Number.isFinite fallback to 0, so malformed values cannot persist NaN. Added adapter regression coverage and pre-push passed.

File: lib/kernel/migrations.js [hunk 9]
```diff
@@ -79,6 +93,14 @@ function buildSchemaMigration(schema = getKernelSchema()) {
...
+		rollback: ['ALTER TABLE kernel_events DROP COLUMN expected_revision;'],
+	};
+}
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/kernel/migrations.js:109]: Fixed in c59444e: rollback no longer uses SQLite DROP COLUMN. The 002 rollback recreates kernel_events without expected_revision, copies preserved columns, renames the table, and recreates the event indexes. Migration tests and pre-push passed.
